### PR TITLE
refactor(demultiplexing api)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

Move the function `get_all_demultiplexed_flow_cell_dirs` from the `DemultiplexingAPI` to the `DemuxPostProcessingAPI`

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/cli/clean.py [hunk 1]
```diff
@@ -449,10 +450,10 @@ def remove_old_flow_cell_run_dirs(context: CGConfig, sequencer: str, days_old: i
 def remove_old_demutliplexed_run_dirs(context: CGConfig, days_old: int, dry_run: bool):
     """Removes flow cells from demultiplexed run directory."""
     status_db: Store = context.status_db
-    demux_api: DemultiplexingAPI = context.demultiplex_api
+    demux_post_processing_api: DemuxPostProcessingAPI = DemuxPostProcessingAPI(config=context)
```
Comment (seallard) [inline cg/cli/clean.py:453]: I think we should stick to the pattern of setting up the `apis` in the context and not instantiate them directly where they are used. Probably out of scope for this PR

**Review by seallard — APPROVED** [review]
Nice! Is the goal to eliminate the `demultiplex_api.py` completely? 

I think we should instantiate the api in the context instead of directly in the command, but out of scope for now!

File: tests/apps/demultiplex/test_demultiplex_api.py
Comment (diitaz93) [inline tests/apps/demultiplex/test_demultiplex_api.py:1]: Is there a reason to leave this file?

**Review by diitaz93 — APPROVED** [review]
Great idea 💯

**Review by islean — APPROVED** [review]
Looks nice!

**Comment by ChrOertlin** [comment]
> Nice! Is the goal to eliminate the `demultiplex_api.py` completely?
> 
> I think we should instantiate the api in the context instead of directly in the commands, but out of scope for this PR!

No, the aim is to decouple the demuxpostprocessing API from the demultiplex api. Currently the demuxpostprocssing api take a directory from the demultiplex api which is also in the config. We can just fetch it from there. Also this the function that gets moved in this PR was shared. However, it makes more sense to be part of the demux post processing API as it gets used here and happens when demultiplexing is finished.

File: tests/meta/demultiplex/test_demux_post_processing.py [hunk 2]
```diff
@@ -567,3 +567,19 @@ def test_post_processing_of_flow_cell(
...
+    # GIVEN a demultiplex flow cell finished output directory that exist
+
+    # GIVEN a demultiplex context
+    demux_api: DemuxPostProcessingAPI = DemuxPostProcessingAPI(config=demultiplex_context)
+
+    # WHEN calling get_all_demultiplexed_flow_cell_dirs
+    demultiplex_flow_cell_dirs: List[Path] = demux_api.get_all_demultiplexed_flow_cell_dirs()
+
+    # THEN the demultiplexed flow cells run directories should be returned
+    assert demultiplex_flow_cell_dirs[0] == demultiplexed_flow_cell_finished_working_directory
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply tests/meta/demultiplex/test_demux_post_processing.py:585]: With manual intervention I think that is possible. But there are checks in place for parsed a valid flow cell dir.

File: cg/cli/clean.py (hunk 1 above)
Comment (ChrOertlin) [reply cg/cli/clean.py:453]: I opened an issue regardin this

File: tests/apps/demultiplex/test_demultiplex_api.py
Comment (ChrOertlin) [reply tests/apps/demultiplex/test_demultiplex_api.py:1]: Nope, removed it

**Comment by ChrOertlin** [comment]
# cg clean remove-old-demultiplexed-run-dirs --dry-run
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Instantiating cgstats api
Set environment to stage
Called undefined __fields__ on HousekeeperAPI, please wrap
Demultiplexing not finished for H27NHDRX3. Skipping removal of the directory.
Demultiplexing not finished for HFFJNALXX. Skipping removal of the directory.
Demultiplexing not finished for H2FYJCCXY. Skipping removal of the directory.
Demultiplexing not finished for H3GVHCCXX. Skipping removal of the directory.
Demultiplexing not finished for HWV5LDRX2. Skipping removal of the directory.
Demultiplexing not finished for HWNLYDSX5. Skipping removal of the directory.
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
Fetching files with tags in [fastq]
Fetching files from version 145558
Fetching files with tags in [fastq]
Fetching files from version 145417
Fetching files with tags in [fastq]
Fetching files from version 145504
Fetching files with tags in [fastq]
Fetching files from version 145035


# cg demutliplex finish all

cg demultiplex finish all
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Instantiating cgstats api
Set environment to stage
Check demuxed flow cell 230503_A00621_0843_BH27NHDRX3
Check if demultiplexing is ready for /home/proj/stage/flowcells/novaseq/runs/230503_A00621_0843_BH27NHDRX3
/home/proj/stage/demultiplexed-runs/230503_A00621_0843_BH27NHDRX3/demuxcomplete.txt
Demultiplex is not ready for 230503_A00621_0843_BH27NHDRX3
Check demuxed flow cell 170120_ST-E00269_0165_BHFFJNALXX
Check if demultiplexing is ready for /home/proj/stage/flowcells/novaseq/runs/170120_ST-E00269_0165_BHFFJNALXX
Could not find logfile: /home/proj/stage/flowcells/novaseq/runs/170120_ST-E00269_0165_BHFFJNALXX/HFFJNALXX_demultiplex.stderr!
Demultiplex is not ready for 170120_ST-E00269_0165_BHFFJNALXX
Check demuxed flow cell 170613_ST-E00266_0216_BH2FYJCCXY
Check if demultiplexing is ready for /home/proj/stage/flowcells/novaseq/runs/170613_ST-E00266_0216_BH2FYJCCXY
Could not find logfile: /home/proj/stage/flowcells/novaseq/runs/170613_ST-E00266_0216_BH2FYJCCXY/H2FYJCCXY_demultiplex.stderr!
Demultiplex is not ready for 170613_ST-E00266_0216_BH2FYJCCXY
Check demuxed flow cell 151125_ST-E00266_0055_BH3GVHCCXX
Check if demultiplexing is ready for /home/proj/stage/flowcells/novaseq/runs/151125_ST-E00266_0055_BH3GVHCCXX
Could not find logfile: /home/proj/stage/flowcells/novaseq/runs/151125_ST-E00266_0055_BH3GVHCCXX/H3GVHCCXX_demultiplex.stderr!
Demultiplex is not ready for 151125_ST-E00266_0055_BH3GVHCCXX
Check demuxed flow cell 221214_A00689_0717_AHWV5LDRX2
Check if demultiplexing is ready for /home/proj/stage/flowcells/novaseq/runs/221214_A00689_0717_AHWV5LDRX2
Could not find logfile: /home/proj/stage/flowcells/novaseq/runs/221214_A00689_0717_AHWV5LDRX2/HWV5LDRX2_demultiplex.stderr!
Demultiplex is not ready for 221214_A00689_0717_AHWV5LDRX2

**Comment by ChrOertlin** [comment]
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs
done.
Log deploy... done.
cg, version 40.9.6

**Pull request merged**